src/chainstate/stacks/db/blocks.rs

```diff
@@ -4126,14 +4154,25 @@ impl StacksChainState {
             "No coinbase transaction".to_string(),
         ))?;
         let miner_auth = coinbase_tx.get_origin();
-        let miner_addr = if mainnet {
-            miner_auth.address_mainnet()
+        let miner_addr = miner_auth.get_address(mainnet);
+
+        let recipient_contract = if epoch_id >= StacksEpochId::Epoch21 {
+            // pay to tx-designated recipient, or if there is none, pay to the origin
+            match coinbase_tx.try_as_coinbase() {
+                Some((_, contract_id_opt)) => contract_id_opt.cloned(),
+                None => None,
+            }
         } else {
-            miner_auth.address_testnet()
+            // pre-2.1, always pay to the origin
+            None
         };
 
+        // N.B. a `MinerPaymentSchedule` that pays to a contract can never be created before 2.1,
+        // per the above check (and moreover, a Stacks block with a pay-to-contract coinbase would
+        // not become valid until after 2.1 activates).
         let miner_reward = MinerPaymentSchedule {
             address: miner_addr,
+            recipient_contract: recipient_contract,
```

Comment:
Can this field instead be `recipient` and _always_ used (i.e., rather than being `Option<QualifiedContractPrincipal>`, it would be `PrincipalData`)?

That way, we don't have a conditional logic field in the database (and in this struct) which requires repeating the check in lines 4159-4167 any time we want to figure out the recipient of the scheduled payment.

Reply:
I'm not sure I understand the question?  We're going to need to gate this either way, since we cannot permit the _possibility_ that the current or future node implementation will allow a pay-to-contract coinbase transaction to be processed.  Representing the recipient contract as a `PrincipalData` instead of `Option<QualifiedContractPrincipal>` does not free you from this.

Reply:
The idea is that the `MinerPaymentSchedule` struct just specifies a recipient principal -- we construct valid MinerPaymentSchedule instances when processing coinbase transactions, and then when processing the scheduled payment, we just debit `recipient`.

This leads to a clearer database schema: to figure out the recipients from the scheduled payments table, you just look at the recipient column, instead of "look at the miner address column and the recipient_contract column, if the second is null, then use the first, otherwise use the second".

Regarding the gates: you would gate the constructor for scheduled payments. If a scheduled payment struct is constructed that is invalid, that's an implementation error, and should panic.

Reply:
Gotcha.  I've updated the code so that:

* The `recipient` column is always set on future DB inserts, but if it's not set, it defaults to `address`
* `recipient` is now a `PrincipalData` in `MinerPaymentSchedule` and `MinerReward`.
* The `TransactionPayload::Coinbase` variant takes an `Option<PrincipalData>`, and will serialize it as a `Value` via the `StacksMessageCodec` implementation for `Value`.
* Test coverage is expanded to verify that a block reward can be sent to either a standard or contract principal address